src/cholesky.jl

```diff
@@ -1,21 +1,50 @@
 # Generic Cholesky decomposition for fixed-size matrices, mostly unrolled
 non_hermitian_error() = throw(LinearAlgebra.PosDefException(-1))
-@inline function LinearAlgebra.cholesky(A::StaticMatrix)
-    ishermitian(A) || non_hermitian_error()
+@inline function LinearAlgebra.cholesky(A::StaticMatrix; check=true)
+    !ishermitian(A) && (check && non_hermitian_error())
     C = _cholesky(Size(A), A)
-    return Cholesky(C, 'U', 0)
+    loc = _checkpsd(C)
+    if check && loc != 0
+        throw(PosDefException(loc))
+    end
+    return Cholesky(C, 'U', loc)
+end
+
+"""
+    _checkpsd(C::StaticMatrix{M,M,T})
+
+Checks of an upper-triangular Cholesky factor `C` was successfully factored.
+If the original matrix `A = C'C` was positive definite, return 0. 
+Otherwise, return the location of the first zero element on the diagonal of `C`.
+
+Used to fill in the `info` field of the `LinearAlgebra.Cholesky` type.
+"""
+function _checkpsd(C::StaticMatrix{M,M,T}) where {M,T}
+    ispsd = abs(prod(diag(C))) > zero(real(eltype(C))) 
+    loc = ispsd ? 0 : findfirst(diag(C) .== zero(eltype(C)))
+    return loc
 end
 
-@inline function LinearAlgebra.cholesky(A::LinearAlgebra.RealHermSymComplexHerm{<:Real, <:StaticMatrix})
+function LinearAlgebra.isposdef(A::StaticMatrix{M,M,T}) where {M,T}
+    C = _cholesky(Size(A), A)
```

Comment:
Shouldn't this check first that `A` is Hermitian? Like here: https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/dense.jl#L70

Reply:
By the way, `isposdef` is defined in `LinearAlgebra` for objects of type `Cholesky`: https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/cholesky.jl#L585 , and it's a different definition.